x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java

```diff
@@ -0,0 +1,112 @@
+import java.util.TreeMap;
+import java.util.concurrent.Executor;
+
+import static org.elasticsearch.xpack.security.support.SecurityMigrations.isEligibleSecurityMigration;
+
+public class SecurityMigrationExecutor extends PersistentTasksExecutor<SecurityMigrationTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(SecurityMigrationExecutor.class);
+    private final SecurityIndexManager securityIndexManager;
+    private final Client client;
+    private final FeatureService featureService;
+    private final ClusterService clusterService;
+    private final TreeMap<Integer, SecurityMigrations.SecurityMigration> migrationByVersion;
+
+    public SecurityMigrationExecutor(
+        String taskName,
+        Executor executor,
+        SecurityIndexManager securityIndexManager,
+        Client client,
+        FeatureService featureService,
+        ClusterService clusterService,
+        TreeMap<Integer, SecurityMigrations.SecurityMigration> migrationByVersion
+    ) {
+        super(taskName, executor);
+        this.securityIndexManager = securityIndexManager;
+        this.client = client;
+        this.featureService = featureService;
+        this.clusterService = clusterService;
+        this.migrationByVersion = migrationByVersion;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, SecurityMigrationTaskParams params, PersistentTaskState state) {
+        applyOutstandingMigrations(task, params.getMigrationVersion(), ActionListener.wrap((res) -> task.markAsCompleted(), (exception) -> {
+            logger.warn("Security migration failed: " + exception);
+            task.markAsFailed(exception);
+        }));
+    }
+
+    private void applyOutstandingMigrations(AllocatedPersistentTask task, int currentMigrationVersion, ActionListener<Void> listener) {
+        if (task.isCancelled()) {
+            listener.onFailure(new TaskCancelledException("Security migration task cancelled"));
+            return;
+        }
+        if (migrationByVersion.lastKey() > currentMigrationVersion) {
+            Map.Entry<Integer, SecurityMigrations.SecurityMigration> migrationEntry = migrationByVersion.higherEntry(
+                currentMigrationVersion
+            );
+            SecurityMigrations.SecurityMigration migration = migrationEntry.getValue();
+
+            // Check if all nodes can support feature and that the cluster is on a compatible mapping version
+            if (isEligibleSecurityMigration(
+                securityIndexManager.indexMappingVersion(),
+                featureService,
+                clusterService.state(),
+                migration
+            )) {
+                migration.migrate(
+                    client,
+                    ActionListener.wrap(
+                        response -> updateMigrationVersion(
+                            migrationEntry.getKey(),
+                            securityIndexManager.getConcreteIndexName(),
```

Comment:
So, it's "reindex this specific concrete index", then "update version on this same specific concrete index" .

Reply:
Yes, I see. The way I think about it is that the a `SecurityMigration` and the `SecurityMigrationExecutor` are specialized implementations that deal with the security index and that the `UpdateIndexMigrationVersionAction` is a generic implementation that can deal with migration statuses of any system index. Passing the index name in the `migrate` method therefore feels redundant. If migrate wasn't part of a security specific interface, then I think passing the index name would make more sense. Does that make sense you think? Happy to discuss more. 

An alternative could be to get the concrete index name by passing the `SecurityIndexManager` to the `migrate` method. I feel like it makes sense to have that part of the interface.

Reply:
The issue I'm trying to cover for with this suggestion is when the `.security` alias points to multiple indices.
There have been support cases. I would like us avoid having to think through "what's the state of the migration if `.security` points to multiple indices".
Related, I have qualms when transport actions try to resolve the `.security` alias. In general, it might need some indices options to go through correctly (because it is a system/hidden/whatever countless flags and options in whatever contexts....mess).
We can bypass all this mess if we refactor the migration to refer to the concrete main security index, the same way that a lot of other security code refers to it, via the SecurityIndexManager.

I can see that, currently, migrations will not be triggered if the .security points to multiple indices, but that seems more like an accident (that ideally we should have a test for).

>An alternative could be to get the concrete index name by passing the SecurityIndexManager to the migrate method. I feel like it makes sense to have that part of the interface.
Yeah, lets go with that.

Reply:
Thanks for explaining. That makes sense! I'll update to use the concrete index name.